libcnb-test/src/macros.rs

```diff
@@ -0,0 +1,48 @@
+/// Asserts that `left` contains `right`.
+///
+/// Commonly used when asserting `pack` output in integration tests. Expands to a [`str::contains`]
+/// call and logs `left` (in unescaped and escaped form) as well as `right` on failure.
+///
+/// # Example
+///
+/// ```
+/// use libcnb_test::assert_contains;
+///
+/// let output = "Hello World!\nHello Integration Test!";
+/// assert_contains!(output, "Integration");
+/// ```
+#[macro_export]
+macro_rules! assert_contains {
+    ($left:expr, $right:expr $(,)?) => {{
+        if !$left.contains($right) {
+            ::std::panic!(
+                r#"assertion failed: `(left contains right)`
+left (unescaped):
+{}
+
+left (escaped): `{:?}`
+right: `{:?}`"#,
```

Comment:
If I edit the tests in `examples/example-02-ruby-sample/tests/integration_test.rs` to make one of the assertions fail, then run `cargo test -- --ignored`, I get the following output:

```
...
test basic ... FAILED

failures:

---- basic stdout ----
thread 'basic' panicked at 'assertion failed: `(left contains right)`
left (unescaped):
20: Pulling from heroku/buildpacks
ea362f368469: Already exists
72f4a9f1dba4: Already exists
3adc8fa488d3: Already exists
49a7135e1277: Already exists
52de67cc0efc: Already exists
0a854087d327: Pulling fs layer
babca8a4e3d2: Pulling fs layer
357fefdf9bc9: Pulling fs layer
d9d5783cd0f5: Pulling fs layer
b59bebfeca61: Pulling fs layer
f09639575729: Pulling fs layer
6000137ea975: Pulling fs layer
7879deff42eb: Pulling fs layer
ad3978c5ff0a: Pulling fs layer
f05ce5f7b606: Pulling fs layer
f09639575729: Waiting
b59bebfeca61: Waiting
7879deff42eb: Waiting
46c9d96303bf: Pulling fs layer
7a5140431caf: Pulling fs layer
c10fd1b9aa75: Pulling fs layer
f05ce5f7b606: Waiting
46c9d96303bf: Waiting
5035dafe4b03: Pulling fs layer
31fdc8b76403: Pulling fs layer
c10fd1b9aa75: Waiting
bcbb9964778f: Pulling fs layer
1d5c43878f1d: Pulling fs layer
be25d4961b5e: Pulling fs layer
dd55db6fa015: Pulling fs layer
bcbb9964778f: Waiting
0a9ff40b2cb3: Pulling fs layer
587b69128b60: Pulling fs layer
31fdc8b76403: Waiting
1d5c43878f1d: Waiting
6a19c4a9cc6d: Pulling fs layer
c40484cf3f96: Pulling fs layer
9215aa0839f1: Pulling fs layer
4f4fb700ef54: Pulling fs layer
587b69128b60: Waiting
6a19c4a9cc6d: Waiting
c40484cf3f96: Waiting
4f4fb700ef54: Waiting
9215aa0839f1: Waiting
357fefdf9bc9: Verifying Checksum
357fefdf9bc9: Download complete
0a854087d327: Verifying Checksum
0a854087d327: Download complete
babca8a4e3d2: Download complete
0a854087d327: Pull complete
babca8a4e3d2: Pull complete
d9d5783cd0f5: Verifying Checksum
d9d5783cd0f5: Download complete
357fefdf9bc9: Pull complete
b59bebfeca61: Verifying Checksum
b59bebfeca61: Download complete
f09639575729: Verifying Checksum
f09639575729: Download complete
d9d5783cd0f5: Pull complete
6000137ea975: Verifying Checksum
6000137ea975: Download complete
b59bebfeca61: Pull complete
f09639575729: Pull complete
6000137ea975: Pull complete
7879deff42eb: Verifying Checksum
7879deff42eb: Download complete
ad3978c5ff0a: Verifying Checksum
ad3978c5ff0a: Download complete
f05ce5f7b606: Verifying Checksum
f05ce5f7b606: Download complete
7879deff42eb: Pull complete
ad3978c5ff0a: Pull complete
7a5140431caf: Verifying Checksum
7a5140431caf: Download complete
f05ce5f7b606: Pull complete
46c9d96303bf: Download complete
c10fd1b9aa75: Verifying Checksum
c10fd1b9aa75: Download complete
46c9d96303bf: Pull complete
7a5140431caf: Pull complete
5035dafe4b03: Verifying Checksum
5035dafe4b03: Download complete
31fdc8b76403: Verifying Checksum
31fdc8b76403: Download complete
c10fd1b9aa75: Pull complete
bcbb9964778f: Verifying Checksum
bcbb9964778f: Download complete
5035dafe4b03: Pull complete
31fdc8b76403: Pull complete
bcbb9964778f: Pull complete
be25d4961b5e: Verifying Checksum
be25d4961b5e: Download complete
dd55db6fa015: Verifying Checksum
dd55db6fa015: Download complete
1d5c43878f1d: Download complete
1d5c43878f1d: Pull complete
0a9ff40b2cb3: Download complete
587b69128b60: Verifying Checksum
587b69128b60: Download complete
be25d4961b5e: Pull complete
6a19c4a9cc6d: Download complete
dd55db6fa015: Pull complete
0a9ff40b2cb3: Pull complete
587b69128b60: Pull complete
6a19c4a9cc6d: Pull complete
c40484cf3f96: Verifying Checksum
c40484cf3f96: Download complete
c40484cf3f96: Pull complete
9215aa0839f1: Download complete
4f4fb700ef54: Download complete
9215aa0839f1: Pull complete
4f4fb700ef54: Pull complete
Digest: sha256:8af1daa20c273a24d6ce382ad1282f469fe74a2b0c910371ea5bc525a88d2d1a
Status: Downloaded newer image for heroku/buildpacks:20
20: Pulling from heroku/pack
ea362f368469: Already exists
72f4a9f1dba4: Already exists
3adc8fa488d3: Already exists
31a064c5f57f: Pulling fs layer
a82f421d55c6: Pulling fs layer
a82f421d55c6: Verifying Checksum
a82f421d55c6: Download complete
31a064c5f57f: Pull complete
a82f421d55c6: Pull complete
Digest: sha256:39c6656aa5eebed8df28c34cd785db7263ba933174d62fe6dcb7bec67e4276a6
Status: Downloaded newer image for heroku/pack:20
===> ANALYZING
Previous image with name "libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw" not found
===> DETECTING
libcnb-examples/ruby 0.1.0
===> RESTORING
===> BUILDING
---> Ruby Buildpack
---> Download and extracting Ruby
---> Installing bundler
Successfully installed bundler-2.3.7
Parsing documentation for bundler-2.3.7
Installing ri documentation for bundler-2.3.7
Done installing documentation for bundler after 0 seconds
1 gem installed
---> Installing gems
Fetching gem metadata from https://rubygems.org/.
Fetching minitest 5.14.4
Using bundler 2.3.7
Installing minitest 5.14.4
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Bundled gems are installed into `/layers/libcnb-examples_ruby/bundler`
===> EXPORTING
Adding layer 'libcnb-examples/ruby:bundler'
Adding layer 'libcnb-examples/ruby:ruby'
Adding 1/1 app layer(s)
Adding layer 'launcher'
Adding layer 'config'
Adding layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw...
*** Images (b13ca57ee98d):
      libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw
Adding cache layer 'libcnb-examples/ruby:bundler'
Successfully built image 'libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw'


left (escaped): `"20: Pulling from heroku/buildpacks\nea362f368469: Already exists\n72f4a9f1dba4: Already exists\n3adc8fa488d3: Already exists\n49a7135e1277: Already exists\n52de67cc0efc: Already exists\n0a854087d327: Pulling fs layer\nbabca8a4e3d2: Pulling fs layer\n357fefdf9bc9: Pulling fs layer\nd9d5783cd0f5: Pulling fs layer\nb59bebfeca61: Pulling fs layer\nf09639575729: Pulling fs layer\n6000137ea975: Pulling fs layer\n7879deff42eb: Pulling fs layer\nad3978c5ff0a: Pulling fs layer\nf05ce5f7b606: Pulling fs layer\nf09639575729: Waiting\nb59bebfeca61: Waiting\n7879deff42eb: Waiting\n46c9d96303bf: Pulling fs layer\n7a5140431caf: Pulling fs layer\nc10fd1b9aa75: Pulling fs layer\nf05ce5f7b606: Waiting\n46c9d96303bf: Waiting\n5035dafe4b03: Pulling fs layer\n31fdc8b76403: Pulling fs layer\nc10fd1b9aa75: Waiting\nbcbb9964778f: Pulling fs layer\n1d5c43878f1d: Pulling fs layer\nbe25d4961b5e: Pulling fs layer\ndd55db6fa015: Pulling fs layer\nbcbb9964778f: Waiting\n0a9ff40b2cb3: Pulling fs layer\n587b69128b60: Pulling fs layer\n31fdc8b76403: Waiting\n1d5c43878f1d: Waiting\n6a19c4a9cc6d: Pulling fs layer\nc40484cf3f96: Pulling fs layer\n9215aa0839f1: Pulling fs layer\n4f4fb700ef54: Pulling fs layer\n587b69128b60: Waiting\n6a19c4a9cc6d: Waiting\nc40484cf3f96: Waiting\n4f4fb700ef54: Waiting\n9215aa0839f1: Waiting\n357fefdf9bc9: Verifying Checksum\n357fefdf9bc9: Download complete\n0a854087d327: Verifying Checksum\n0a854087d327: Download complete\nbabca8a4e3d2: Download complete\n0a854087d327: Pull complete\nbabca8a4e3d2: Pull complete\nd9d5783cd0f5: Verifying Checksum\nd9d5783cd0f5: Download complete\n357fefdf9bc9: Pull complete\nb59bebfeca61: Verifying Checksum\nb59bebfeca61: Download complete\nf09639575729: Verifying Checksum\nf09639575729: Download complete\nd9d5783cd0f5: Pull complete\n6000137ea975: Verifying Checksum\n6000137ea975: Download complete\nb59bebfeca61: Pull complete\nf09639575729: Pull complete\n6000137ea975: Pull complete\n7879deff42eb: Verifying Checksum\n7879deff42eb: Download complete\nad3978c5ff0a: Verifying Checksum\nad3978c5ff0a: Download complete\nf05ce5f7b606: Verifying Checksum\nf05ce5f7b606: Download complete\n7879deff42eb: Pull complete\nad3978c5ff0a: Pull complete\n7a5140431caf: Verifying Checksum\n7a5140431caf: Download complete\nf05ce5f7b606: Pull complete\n46c9d96303bf: Download complete\nc10fd1b9aa75: Verifying Checksum\nc10fd1b9aa75: Download complete\n46c9d96303bf: Pull complete\n7a5140431caf: Pull complete\n5035dafe4b03: Verifying Checksum\n5035dafe4b03: Download complete\n31fdc8b76403: Verifying Checksum\n31fdc8b76403: Download complete\nc10fd1b9aa75: Pull complete\nbcbb9964778f: Verifying Checksum\nbcbb9964778f: Download complete\n5035dafe4b03: Pull complete\n31fdc8b76403: Pull complete\nbcbb9964778f: Pull complete\nbe25d4961b5e: Verifying Checksum\nbe25d4961b5e: Download complete\ndd55db6fa015: Verifying Checksum\ndd55db6fa015: Download complete\n1d5c43878f1d: Download complete\n1d5c43878f1d: Pull complete\n0a9ff40b2cb3: Download complete\n587b69128b60: Verifying Checksum\n587b69128b60: Download complete\nbe25d4961b5e: Pull complete\n6a19c4a9cc6d: Download complete\ndd55db6fa015: Pull complete\n0a9ff40b2cb3: Pull complete\n587b69128b60: Pull complete\n6a19c4a9cc6d: Pull complete\nc40484cf3f96: Verifying Checksum\nc40484cf3f96: Download complete\nc40484cf3f96: Pull complete\n9215aa0839f1: Download complete\n4f4fb700ef54: Download complete\n9215aa0839f1: Pull complete\n4f4fb700ef54: Pull complete\nDigest: sha256:8af1daa20c273a24d6ce382ad1282f469fe74a2b0c910371ea5bc525a88d2d1a\nStatus: Downloaded newer image for heroku/buildpacks:20\n20: Pulling from heroku/pack\nea362f368469: Already exists\n72f4a9f1dba4: Already exists\n3adc8fa488d3: Already exists\n31a064c5f57f: Pulling fs layer\na82f421d55c6: Pulling fs layer\na82f421d55c6: Verifying Checksum\na82f421d55c6: Download complete\n31a064c5f57f: Pull complete\na82f421d55c6: Pull complete\nDigest: sha256:39c6656aa5eebed8df28c34cd785db7263ba933174d62fe6dcb7bec67e4276a6\nStatus: Downloaded newer image for heroku/pack:20\n===> ANALYZING\nPrevious image with name \"libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw\" not found\n===> DETECTING\nlibcnb-examples/ruby 0.1.0\n===> RESTORING\n===> BUILDING\n---> Ruby Buildpack\n---> Download and extracting Ruby\n---> Installing bundler\nSuccessfully installed bundler-2.3.7\nParsing documentation for bundler-2.3.7\nInstalling ri documentation for bundler-2.3.7\nDone installing documentation for bundler after 0 seconds\n1 gem installed\n---> Installing gems\nFetching gem metadata from https://rubygems.org/.\nFetching minitest 5.14.4\nUsing bundler 2.3.7\nInstalling minitest 5.14.4\nBundle complete! 1 Gemfile dependency, 2 gems now installed.\nBundled gems are installed into `/layers/libcnb-examples_ruby/bundler`\n===> EXPORTING\nAdding layer 'libcnb-examples/ruby:bundler'\nAdding layer 'libcnb-examples/ruby:ruby'\nAdding 1/1 app layer(s)\nAdding layer 'launcher'\nAdding layer 'config'\nAdding layer 'process-types'\nAdding label 'io.buildpacks.lifecycle.metadata'\nAdding label 'io.buildpacks.build.metadata'\nAdding label 'io.buildpacks.project.metadata'\nSetting default process type 'web'\nSaving libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw...\n*** Images (b13ca57ee98d):\n      libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw\nAdding cache layer 'libcnb-examples/ruby:bundler'\nSuccessfully built image 'libcnbtest_qidplmbovdmxgtyijmakcpprwdwohw'\n"`
right: `"AAA"`', examples/example-02-ruby-sample/tests/integration_test.rs:23:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    basic

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 44.02s
```

I find this output quite hard to read. Whilst there is an unescaped version of the console output, it requires scrolling up to see as it isn't the last part of the output, and it duplicates the escaped version.

The above is admittedly not helped by the docker pull output, which could be reduced by implementing #306 short term (and then buildpacks/pack#1368 + [W-9864812](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000004ytVbYAI/view) longer term). However that change won't help CI, where the images will always be pulled. ie: If the first integration test to run in CI fails due to an `assert_contains` then it will always include hard to read output.

I presume this is partly modelled from:
https://github.com/rust-lang/rust/blob/25b21a1d160013b2d58c98d0337c75a721317160/library/core/src/panicking.rs#L219-L230

I can understand why both the escaped + unescaped output is included (in case there's a difference between the two that's not easily identifiable from the escaped version), however I wonder if this perhaps worsens the UX in the common case to order to improve the UX in an uncommon case?

ie: Should we perhaps instead print only the unescaped versions of both?

Reply:
> I find this output quite hard to read.

The code block doesn't quite do the escaped version justice (since when rendered in this thread it doesn't wrap the lines) - here's what it looks like in my VSCode console:

<img width="575" alt="example assert_contains failure console output" src="https://user-images.githubusercontent.com/501702/154043898-9a407488-b922-40bc-8ea2-c985bb2b10bc.png">

Reply:
I see your point, but I wonder what the story for the "less common" case would be then. I've experienced issues with whitespace in my buildpack test code before - it's not super rare. If we'd remove the unescaped output, we'd need an easy way to enable it somehow. Otherwise DX is also not great IMO when I have to start reading docs, maybe set env vars or even write custom debug code just to figure out why my assertions fails.

I agree that in narrow terminals this is a sea of characters, but it's not like it's always there - only when a test fails which is hopefully not the standard case. :)

For me, the benefits of having the verbose output outweigh the downsides.

Reply:
That said, I'm happy to perhaps just see how this works in practice, once we start writing some tests?

Reply:
Sounds good to me. :) 👍🏻